draft-rha-jose-hpke-encrypt.md

```diff
@@ -241,18 +244,55 @@ In Integrated Encryption mode:
 
 ~~~
-eyJhbGciOiJIUEtFLUJhc2UtUDI1Ni1TSEEyNTYtQUVTMTI4R0NNIiwiZXBrIjp7Imt0eSI6IkVLIiwiZWsiOiJCQU9TeWV3M05JLUkwNEd2WU1MT3Y0cDBEVG5WMWZjWnBFVW10dGs0YkRTdDAtakxzY0FDN3h3MjdORTFHZ0VuMUgtM3ZXSFA5eW1BOHl4aFRmVDBkYjQifX0...afBw3T1hUNjci4qq3ZZ-9KxnttB0iCEO_GUqbIStqYqB5DgRDpyYSuvoH1mMA31qKPqB41ld5mSP34yUys6WJM7nstDJ1-4nqUdhRpgfkGTECA.
+eyJhbGciOiJkaXIiLCJlbmMiOiJIUEtFLUJhc2UtUDI1Ni1TSEEyNTYtQUVTMTI4R0NNIiwiZXBrIjp7Imt0eSI6IkVLIiwiZWsiOiJCR05ranp0MDc2YnNSR2o3OGFYNUF6VF9IRU9KQmJZOXEyWm9fNWU3dGJLMGFQcXU0ZVQxV0kxNmp2UmxacXBNeXFaZlAtUndSNEp3dGF6XzhVOXRoWEEifX0...DG3qygxcMHw3iACy5mX_T4N4EqWc03W0nkTHjMJsC4nb6JS6vVj6wTGdlr5TOSr0ykaoyzpePXEvEyHhvpUwCyQQr6kbGlGuZsrJdUbZ728vmA.
+~~~
+{: #integrated-encryption-compact align="left" title="Integrated Encryption with Compact Serialization"}
+
+In the above example, the JWE Protected Header value is: 
+
+~~~
+{
+  "alg": ""HPKE-IntEnc",
+  "enc": "HPKE-Base-P256-SHA256-AES128GCM",
+  "epk": {
+    "kty": "EK",
+    "ek": "BGNkjzt076bsRGj78aX5AzT_HEOJBbY9q2Zo_5e7tbK0aPqu4eT1WI16jvRlZqpMyqZfP-RwR4Jwtaz_8U9thXA"
+  }
+}
 ~~~
-{: #compact-example align="left" title="Integrated Encryption with Compact Serialization"}
+
+~~~
+{
+    "protected": "eyJhbGciOiJkaXIiLCJlbmMiOiJIUEtFLUJhc2UtUDI1Ni1TSEEyNTYtQUVTMTI4R0NNIiwiZXBrIjp7Imt0eSI6IkVLIiwiZWsiOiJCTzRFbGZXd0xKRDZWcERza3c5LWxWMm9OMDJ2U1FKTW55ZHk3enhvSVlKZ1kydk9taE44Q1BqSHdRM1NONkhTcnNHNElacVpHVUR3dExKZjBoeHFTWGsifX0",
+    "ciphertext": "1ATsw0jshqPrv8CFcm9Rem9Wfi1Ygv30sozlRTtNNzcaaZ828GqP0AXtqQ1Msv8YXI9XZqh81MK3QnlZ7pOBC1BP7j00J1rrHujdb3zvnOpmJg"
+}
+~~~
+{: #integrated-encryption-json align="left" title="Integrated Encryption with JSON Serialization"}
+
+In the above example, the JWE Protected Header value is: 
+
+~~~
+{
+  "alg": ""HPKE-IntEnc",
+  "enc": "HPKE-Base-P256-SHA256-AES128GCM",
+  "epk": {
+    "kty": "EK",
+    "ek": "BGNkjzt076bsRGj78aX5AzT_HEOJBbY9q2Zo_5e7tbK0aPqu4eT1WI16jvRlZqpMyqZfP-RwR4Jwtaz_8U9thXA"
+  }
+}
+~~~
+
+When Integrated Encryption is represented in General JSON Serialization, `iv`, `tag` and `aad` must not be present, as they are always empty.
 
 ### Key Encryption
 
-This mode supports more than one recipient and JSON serializations.
+This mode supports more than one recipient.
 
 HPKE is used to encrypt the	Content Encryption Key (CEK), and the resulting ciphertext is included in the JWE ciphertext. 
 The plaintext will be encrypted using the CEK as explained in Step 15 of Section 5.1 of {{RFC7516}}.
 
-The sender MUST place the 'epk' and 'alg' parameters in the per-recipient unprotected header to indicate the use of HPKE. The "enc" (Encryption Algorithm) Header Parameter MUST be present in the JWE Protected Header.	 		
+The sender MUST place the 'epk' and 'alg' parameters in the per-recipient unprotected header, or shared unprotected header to indicate the use of HPKE. 
+The "enc" (Encryption Algorithm) Header Parameter MUST be present in the JWE Protected Header.	 		
```

Comment:
Just because this mode supports more recipients does not mean it always will be used with more than 1 recipient.

Because of this you cannot say that epk and alg MUST be placed in {enter location}.

The only requirements that should be present are for when something MUST be integrity protected; therefore, it MUST occur only within the JWE Protected Header.

That's it, no more needs to be said.

Reply:
I re-phrased the text to address your comment:
When there are multiple recipients, the sender MUST place the 'epk' and 'alg' parameters in the per-recipient unprotected header or shared unprotected header to indicate the use of HPKE. In this case, the 'enc' (Encryption Algorithm) Header Parameter MUST be present in the JWE Protected Header.

Reply:
- does "enc" actually have to be protected?
- when there are multiple recipients but they all use the same "alg" value, we can't place that "alg" parameter in a protected header?
- when there is one recipient epk can't be protected?
- I could probably go on...

This entire section is inconsistent and redundant with what it is implied from how the JWE serialization formats, JOSE Header structures, and their processing, already function and are defined.

Reply:
"enc" has to be protected, it is discussed https://datatracker.ietf.org/doc/html/rfc7516#section-7.2.1 that says "This  primarily means that the "enc" (encryption algorithm) Header Parameter value in the JOSE Header for each recipient and any parameters of that algorithm MUST be the same."

However, I agree that "alg" can also be placed in the protected header and "epk" can be protected in case of one receipt but the proposed text is specific to multiple recipients. I will modify the text as follows: 

When there are multiple recipients, the sender MUST place the 'epk' parameter in the per-recipient unprotected header to indicate the use of HPKE. In this case, the 'enc' (Encryption Algorithm) Header Parameter MUST be a content encryption algorithm from JSON Web Signature and Encryption Algorithms in {{JOSE-IANA}}, and it MUST be present in the JWE Protected Header.

Reply:
> "enc" has to be protected, it is discussed [datatracker.ietf.org/doc/html/rfc7516#section-7.2.1](https://datatracker.ietf.org/doc/html/rfc7516#section-7.2.1) that says "This primarily means that the "enc" (encryption algorithm) Header Parameter value in the JOSE Header for each recipient and any parameters of that algorithm MUST be the same."

I don't feel that i agree with one leading to the other. Yes, the "enc" values need to be the same for every recipient, clearly, that's inherent to the general JWE format, but that doesn't mean it has to be protected in a produced JWE. Neither does it mean that we have to re-iterate what's already implicit about JWE formats.

Reply:
"enc" can be protected or shared unprotected, we want to avoid "shared unprotected".  The integrity protected 'enc' provides protection against an attacker who manipulates the encryption algorithm in the 'enc' parameter. The attack is discussed in https://datatracker.ietf.org/doc/draft-ietf-lamps-cms-cek-hkdf-sha256/.

Reply:
Than that's worth calling out in the draft.

Reply:
Yes, added text.